### PR TITLE
Integration of MLSL library

### PR DESCRIPTION
**Pull request by AlekseyMarchuk**

### CODE REVIEW

File: doc/examples/mnist_mlp/CMakeLists.txt [hunk 1]
```diff
@@ -18,12 +18,9 @@ add_executable(mnist_mlp mnist_loader.cpp mnist_mlp.cpp)
...
 if (NGRAPH_DISTRIBUTED_ENABLE)
-    find_package(MPI REQUIRED)
     add_definitions(-DNGRAPH_DISTRIBUTED)
-    include_directories(SYSTEM ${MPI_C_INCLUDE_PATH} ${MPI_CXX_INCLUDE_PATH})
-    link_directories(${MPI_C_LIBRARIES} ${MPI_CXX_LIBRARIES})
-    link_libraries(${MPI_CXX_LIBRARIES})
     add_executable(dist_mnist_mlp mnist_loader.cpp dist_mnist_mlp.cpp)
-    add_dependencies(dist_mnist_mlp ngraph cpu_backend)
-    target_link_libraries(dist_mnist_mlp ngraph cpu_backend)
+    add_dependencies(dist_mnist_mlp ngraph cpu_backend libmlsl)
```
Comment (rkimballn1) [inline doc/examples/mnist_mlp/CMakeLists.txt]: This is not needed. The target_link_libraries automatically adds dependencies for libraries linked in.

File: cmake/external_mlsl.cmake [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    INSTALL_DIR "${EXTERNAL_PROJECTS_ROOT}/MLSL/install"
+    EXCLUDE_FROM_ALL TRUE
+    )
+
+ExternalProject_Get_Property(MLSL SOURCE_DIR)
+ExternalProject_Get_Property(MLSL INSTALL_DIR)
+add_library(libmlsl INTERFACE)
+target_include_directories(libmlsl SYSTEM INTERFACE ${SOURCE_DIR}/include)
+target_link_libraries(libmlsl INTERFACE ${INSTALL_DIR}/intel64/lib/libmlsl.so)
+add_dependencies(libmlsl MLSL)
```
Comment (rkimballn1) [inline cmake/external_mlsl.cmake]: Do we need to add an install for this library so when we do `make install`? Do a make install and verify that the libmlsl is there.

File: src/ngraph/codegen/CMakeLists.txt [hunk 3]
```diff
@@ -74,4 +70,9 @@ set_target_properties(codegen PROPERTIES VERSION ${NGRAPH_VERSION} SOVERSION ${N
 add_dependencies(codegen header_resource libmkldnn libeigen)
 target_include_directories(codegen SYSTEM PUBLIC ${CMAKE_BINARY_DIR})
 target_link_libraries(codegen PRIVATE libllvm ngraph)
+if(NGRAPH_DISTRIBUTED_ENABLE)
+    add_dependencies(codegen libmlsl)
+    target_include_directories(codegen SYSTEM PUBLIC libmlsl)
+    target_link_libraries(codegen PUBLIC libmlsl)
```
Comment (rkimballn1) [inline src/ngraph/codegen/CMakeLists.txt]: I don't think we should link libmlsl to codegen. It does not use the library.

File: src/ngraph/runtime/cpu/CMakeLists.txt [hunk 4]
```diff
@@ -157,5 +149,12 @@ if (NGRAPH_CPU_ENABLE)
     target_include_directories(cpu_backend SYSTEM PUBLIC libmkldnn)
     set_target_properties(cpu_backend PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${NGRAPH_BUILD_DIR})
 
+    if(NGRAPH_DISTRIBUTED_ENABLE)
+        add_definitions(-DNGRAPH_DISTRIBUTED)
+        add_dependencies(cpu_backend libmlsl)
```
Comment (rkimballn1) [inline src/ngraph/runtime/cpu/CMakeLists.txt]: not needed

File: src/ngraph/runtime/cpu/CMakeLists.txt [hunk 5]
```diff
@@ -157,5 +149,12 @@ if (NGRAPH_CPU_ENABLE)
...
+        target_include_directories(cpu_backend SYSTEM PUBLIC libmlsl)
```
Comment (rkimballn1) [inline src/ngraph/runtime/cpu/CMakeLists.txt]: Can we make both the include_directories and link_libraries PRIVATE? The less we expose the better.

File: doc/examples/mnist_mlp/dist_mnist_mlp.cpp [hunk 6]
```diff
@@ -107,9 +107,9 @@ float test_accuracy(MNistDataLoader& loader,
            static_cast<float>(sample_count);
 }
 
-int main(int argc, const char* argv[])
+int main(int argc, char* argv[])
 {
-    MPI::Init();
+    MLSL::Environment::GetEnv().Init(&argc, &argv);
```
Comment (rkimballn1) [inline doc/examples/mnist_mlp/dist_mnist_mlp.cpp]: This looks like a classic case where RAII would be a better solution
https://en.cppreference.com/w/cpp/language/raii

File: src/ngraph/runtime/interpreter/CMakeLists.txt [hunk 7]
```diff
@@ -14,20 +14,19 @@
...
 if (NGRAPH_INTERPRETER_ENABLE)
     add_library(interpreter_backend SHARED int_backend.cpp)
     set_target_properties(interpreter_backend PROPERTIES VERSION ${NGRAPH_VERSION} SOVERSION ${NGRAPH_API_VERSION})
     target_link_libraries(interpreter_backend PUBLIC ngraph)
     set_target_properties(interpreter_backend PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${NGRAPH_BUILD_DIR})
 
+if(NGRAPH_DISTRIBUTED_ENABLE)
+    add_definitions(-DNGRAPH_DISTRIBUTED)
+    add_dependencies(interpreter_backend libmlsl)
+    target_include_directories(interpreter_backend SYSTEM PUBLIC libmlsl)
```
Comment (rkimballn1) [inline src/ngraph/runtime/interpreter/CMakeLists.txt]: If we do decide that PRIVATE works then make sure you catch all cases.

File: src/ngraph/runtime/cpu/CMakeLists.txt [hunk 8]
```diff
@@ -157,5 +149,12 @@ if (NGRAPH_CPU_ENABLE)
     target_include_directories(cpu_backend SYSTEM PUBLIC libmkldnn)
     set_target_properties(cpu_backend PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${NGRAPH_BUILD_DIR})
 
+    if(NGRAPH_DISTRIBUTED_ENABLE)
+        add_definitions(-DNGRAPH_DISTRIBUTED)
```
Comment (rkimballn1) [inline src/ngraph/runtime/cpu/CMakeLists.txt]: Use `target_compile_definitions` instead. It is cleaner
https://cmake.org/cmake/help/v3.0/command/target_compile_definitions.html
This is in other parts of the cmake too. I've been meaning to address it...

**Comment by sfvaroglu** [comment]
This PR will break nGraph-MXNet distributed training, since ngraph kvstore uses MPI calls. Have you tested with the bridge?

**Comment by indie** [comment]
@wenzhe-nrv asked me to assign this to myself; probably since some of the changes include changes to the code on the `/doc/examples/mnist_mlp` example, the docs [here](http://ngraph.nervanasys.com/docs/latest/howto/execute.html) and [here](http://ngraph.nervanasys.com/docs/latest/howto/execute.html) will need reviewed

File: src/ngraph/codegen/CMakeLists.txt (hunk 3 above)
Comment (AlekseyMarchuk) [reply src/ngraph/codegen/CMakeLists.txt]: Removed

File: src/ngraph/runtime/cpu/CMakeLists.txt (hunk 4 above)
Comment (AlekseyMarchuk) [reply src/ngraph/runtime/cpu/CMakeLists.txt]: Removed

File: src/ngraph/runtime/cpu/CMakeLists.txt (hunk 5 above)
Comment (AlekseyMarchuk) [reply src/ngraph/runtime/cpu/CMakeLists.txt]: I've changed PUBLIC to PRIVATE, thanks for the comment

File: doc/examples/mnist_mlp/dist_mnist_mlp.cpp (hunk 6 above)
Comment (AlekseyMarchuk) [reply doc/examples/mnist_mlp/dist_mnist_mlp.cpp]: Good idea, thank you

File: src/ngraph/runtime/interpreter/CMakeLists.txt (hunk 7 above)
Comment (AlekseyMarchuk) [reply src/ngraph/runtime/interpreter/CMakeLists.txt]: Changed to PRIVATE

File: src/ngraph/runtime/cpu/CMakeLists.txt (hunk 8 above)
Comment (AlekseyMarchuk) [reply src/ngraph/runtime/cpu/CMakeLists.txt]: Done

**Comment by AlekseyMarchuk** [comment]
Distributed training in the nGraph-MXNet bridge  will be updated to work with MLSL right after merge of this PR

**Comment by sfvaroglu** [comment]
Please have a WIP PR in place so that it can go in right after this one.

File: src/ngraph/runtime/cpu/CMakeLists.txt (hunk 4 above)
Comment (wenzhe-nrv) [reply src/ngraph/runtime/cpu/CMakeLists.txt]: @AlekseyMarchuk FYI, there's no need to reply to the comment about the code you modified based on the review. Github will hide this after your commit so reviewers know you have made the change. :)

File: cmake/external_mlsl.cmake (hunk 2 above)
Comment (AlekseyMarchuk) [reply cmake/external_mlsl.cmake]: make install is called during "make all" invocation. I've added additional cmake install commands to copy .so and header to the common install destination. I've tried to install MLSL to CMAKE_INSTALL_PREFIX directly, but MLSL install target cleans target directory

**Comment by AlekseyMarchuk** [comment]
PR with update of  nGraph-MXNet bridge
https://github.com/NervanaSystems/ngraph-mxnet/pull/388

**Comment by sfvaroglu** [comment]
@lnguyen-nvn This PR adds support to replace OpenMPI with Intel MPI. Could you work on adding support for Intel MPI in CI (both ngraph and ngraph-mxnet)? @AlekseyMarchuk can help with Intel MPI version, how to get it, etc.

**Comment by wenzhe-nrv** [comment]
@AlekseyMarchuk CI failed somehow to fetch the code (I think the current is a fork in your repo). Could you please push this code to a branch in nGraph repo and assign only @sfvaroglu and me as reviewer since both @rkimballn1 and @jbobba both approved the PR? And close this PR after that. thanks.

**Comment by AlekseyMarchuk** [comment]
Created new PR https://github.com/NervanaSystems/ngraph/pull/1520

**Pull request closed without merging**